### PR TITLE
Leaderboards worker

### PR DESCRIPTION
**Pull request by Andrei-Dolgolev**
Add leaderboards worker wich will update all chain only leaderboard automaticly.

```
leaderboards-generator leaderboards-generate --query-api-access-token <moonstream-token>
```

Adding new env variable:
`MOONSTREAM_LEADERBOARD_GENERATOR_JOURNAL_ID`
`MOONSTREAM_API_URL`
`MOONSTREAM_ENGINE_URL`

Structure of leaderboard journal entry:

```
{
    "id": "744c1bd9-8715-4059-9bb9-7d674997085f",
    "journal_url": "https://spire.bugout.dev/journals/8967ed2b-d58c-41c8-bddc-9b0d3e73f64e",
    "content_url": null,
    "title": "Heroes leaderboard CG",
    "content": "{\"leaderboard_id\":\"aeb88f56-6aca-4cb1-9dfd-daeff2c8262c\",\"query_name\":\"leaderboard_aeb88f56_6aca_4cb1_9dfd_daeff2c8262c\",\"params\":{}, \"normalize_addresses\": false}",
    "tags": [
        "leaderboard",
        "leaderboard_id:aeb88f56-6aca-4cb1-9dfd-daeff2c8262c"
    ],
    "created_at": "2023-07-26T20:00:23.499555+00:00",
    "updated_at": "2023-07-26T20:32:30.709585+00:00",
    "context_url": null,
    "context_type": "bugout",
    "context_id": null,
    "locked_by": null
}
```
Content
```
{
    "leaderboard_id": "aeb88f56-6aca-4cb1-9dfd-daeff2c8262c",
    "query_name": "leaderboard_aeb88f56_6aca_4cb1_9dfd_daeff2c8262c",
    "params": {},
    "normalize_addresses": false
}
```

<moonstream-token> - must hold that queries and have permissins to push to leaderboard

### CODE REVIEW

File: crawlers/deploy/leaderboards-worker.service [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+Description=Runs leaderboards generator worker
+After=network.target
+
+[Service]
+Type=oneshot
+WorkingDirectory=/home/ubuntu/moonstream/crawlers/mooncrawl
+EnvironmentFile=/home/ubuntu/moonstream-secrets/app.env
+ExecStart=/home/ubuntu/moonstream-env/bin/python -m mooncrawl.leaderboards_generator.cli leaderboards-generate --query-api-access-token "${MOONSTREAM_PUBLIC_QUERIES_DATA_ACCESS_TOKEN}"
+CPUWeight=60
+SyslogIdentifier=leaderboards-generator
```
Comment (kompotkot) [inline crawlers/deploy/leaderboards-worker.service:11]: Should be as name `leaderboards-worker`

File: crawlers/mooncrawl/mooncrawl/leaderboards_generator/cli.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    """
+
+    ### get leaderboard journal
+
+    query = "#leaderboard"
+
+    if args.leaderboard_id:
+        query += f" #cleaderboard_id:{args.leaderboard_id}"
+
+    leaderboards = bc.search(
```
Comment (kompotkot) [inline crawlers/mooncrawl/mooncrawl/leaderboards_generator/cli.py:39]: I suggest it to put in try except block with return from function so we will handle errors if spire down and do not respond correctly

File: crawlers/mooncrawl/mooncrawl/leaderboards_generator/cli.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+    Get query from journal and push results to leaderboard API.
+    """
+
+    ### get leaderboard journal
+
+    query = "#leaderboard"
+
+    if args.leaderboard_id:
+        query += f" #cleaderboard_id:{args.leaderboard_id}"
```
Comment (kompotkot) [inline crawlers/mooncrawl/mooncrawl/leaderboards_generator/cli.py:37]: Does it typo `" #cl...` - space

File: crawlers/mooncrawl/mooncrawl/leaderboards_generator/cli.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            f"Processing leaderboard: {leaderboard.title} with id: {[tag for tag in leaderboard.tags if tag.startswith('leaderboard_id')]}"
+        )
+
+        if leaderboard.content is None:
+            continue
+
+        try:
+            leaderboard_data = json.loads(leaderboard.content)
+        except json.JSONDecodeError:
+            logger.error(
```
Comment (kompotkot) [inline crawlers/mooncrawl/mooncrawl/leaderboards_generator/cli.py:63]: I sugest also to print entry_id for debug

File: crawlers/mooncrawl/mooncrawl/leaderboards_generator/cli.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            f"Successfully pushed results to leaderboard {info['id']}: {info['title']}"
+        )
+        logger.info(
+            f"can be check on:{args.engine_api}/leaderboard/?leaderboard_id={leaderboard_id}"
+        )
+
+
+def main():
+    """
+    Generates an argument parser for the "autocorns judge" command.
```
Comment (kompotkot) [inline crawlers/mooncrawl/mooncrawl/leaderboards_generator/cli.py:148]: Description should represent correct info :)

File: crawlers/mooncrawl/mooncrawl/leaderboards_generator/utils.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+
+import requests  # type: ignore
+
+
+logging.basicConfig()
+logger = logging.getLogger(__name__)
+
+
+def get_results_for_moonstream_query(
```
Comment (kompotkot) [inline crawlers/mooncrawl/mooncrawl/leaderboards_generator/utils.py:16]: This handles POST request, I sugest to use name: `update_results...` or so

Comment (kompotkot) [reply crawlers/mooncrawl/mooncrawl/leaderboards_generator/utils.py:16]: Not sure about this, anyway better to clarify at least in function description.

File: crawlers/mooncrawl/mooncrawl/leaderboards_generator/utils.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    moonstream_access_token: str,
+    query_name: str,
+    params: Dict[str, Any],
+    api_url: str = "https://api.moonstream.to",
```
Comment (kompotkot) [inline crawlers/mooncrawl/mooncrawl/leaderboards_generator/utils.py:20]: Better use default value from `settings.py`

**Review by kompotkot — COMMENTED** [review]
smc

**Comment by kompotkot** [comment]
@bugout-dev check
- [ ] Create journal under admin user and env variable `MOONSTREAM_LEADERBOARD_GENERATOR_JOURNAL_ID`

File: crawlers/mooncrawl/mooncrawl/leaderboards_generator/utils.py (hunk 6 above)
Comment (Andrei-Dolgolev) [reply crawlers/mooncrawl/mooncrawl/leaderboards_generator/utils.py:18]: Added in description.

**Review by kompotkot — APPROVED** [review]
lg

**Pull request merged**